### ISSUE TITLE
Reproducing the reported numbers and the total running time

### ISSUE THREAD

**Issue by thechargedneutron** [yifan123/flow_grpo#22]
Hi,

Thanks for the excellent work. I am trying to run the code to reproduce the results. Are there any instructions to compare the reported numbers? I see some discussions in #17, but I couldn't follow it fully. Can you please provide a code to reproduce the numbers in the paper?

Also, what is the total running time with your given parameters? I am running the code, and it's been around 48 hours on 1 node, so want to confirm the total running time.

**Comment by thechargedneutron** [comment]
Thanks for the reply! I will try the repo mentioned by you to evaluate the performance. Also, in the logging, `kl` is always zero. Is that expected? Or am I missing something? I am running the default give code without any changes.

<img width="1354" alt="Image" src="https://github.com/user-attachments/assets/dde815da-903c-4d2d-bde4-06a05b61fa34" />

**Comment by yifan123** [comment]
Apologies for the confusion. The `kl_loss` is the standard KL divergence used during training, while `kl_abs` and `kl` represent the KL divergence used as part of the KL reward, i.e., treating KL as a component of the reward. In my final hyperparameters, the KL reward was not used, so it was not calculated here and is displayed as the default value of 0.. You can ignore those and just focus on `kl_loss`.

**Comment by thechargedneutron** [comment]
Hi @yifan123 , thanks for the reply. I have a few more questions regarding multi-node training.

I tried running the code with 2 and 3 nodes. I am not able to see a speedup and faster epoch iteration. Both the 2-node and 3-node variant has sampling and training taking the same time.

For 3-nodes
```
Epoch 2: sampling:  17%|█▋        | 4/24 [01:33<08:58, 26.91s/it]
```
and for 2-nodes
```
Epoch 2: sampling:  17%|█▋        | 4/24 [01:39<09:46, 29.30s/it]
```

And the same for training. Is this expected? I suspect the code is not taking advantage of multi-node and repeating the calculations. Can you please check this?

Also, you mentioned requiring 10 days, do you run the whole, 100000 epochs or just long enough to get good score? Can you let me know approx time it takes for 1 epoch?

**Comment by yifan123** [comment]
You need to modify `config.sample.num_batches_per_epoch = 12` in [here](https://github.com/yifan123/flow_grpo/blob/main/config/dgx.py#L42C7-L42C37). This parameter controls how many data collection steps are performed per epoch. If your setup uses more nodes, you should reduce this value. For example, with 3 nodes, you should set `config.sample.num_batches_per_epoch = 4` (12/3).

The value `100000` epochs is just a large placeholder and will not actually be reached in practice. One epoch takes approximately 0.8 A800 GPU hours. So for a 2-node setup, that's roughly 0.05 hours per epoch.

**Comment by yifan123** [comment]
Remember to update the `address` at [line 19 of  `main.sh`, `main1.sh`,`main2.sh`](https://github.com/yifan123/flow_grpo/blob/main/scripts/multi_node/main2.sh#L19) to the actual IP address of your machine.

**Comment by yifan123** [comment]
1. Yes, these hyperparameters yield better results.  

2. With this KL setting, I trained for roughly 2500 epochs to achieve the final geneval outcome.  

3. Note that my reward_server [https://github.com/yifan123/reward-server/blob/main/gunicorn.conf.py#L28](https://github.com/yifan123/reward-server/blob/main/gunicorn.conf.py#L28) runs locally, so each node can only fetch its own geneval reward. The training code also points to this local IP: [https://github.com/yifan123/flow_grpo/blob/main/flow_grpo/rewards.py#L169](https://github.com/yifan123/flow_grpo/blob/main/flow_grpo/rewards.py#L169).

**Comment by yifan123** [comment]
I use `config.sample.num_batches_per_epoch=4` in 3 nodes with 24 GPUs.

**Comment by thechargedneutron** [comment]
I am revisiting this issue since I am not able to reproduce the exact results. I trained with 3 nodes for roughly 3 days and the performance only reaches around 74% in geneval. The training reward curve looks like this:

<img width="421" alt="Image" src="https://github.com/user-attachments/assets/c643b242-5aee-40cf-936e-9588883d2de5" />

To confirm again, are these the exact parameters that you use?

```
    config.resolution = 512
    config.sample.train_batch_size = 6
    config.sample.num_image_per_prompt = 24
    config.sample.num_batches_per_epoch = 4 # based on our discussion above
    config.sample.test_batch_size = 6 # This bs is a special design, the test set has a total of 2212, to make gpu_num*bs*n as close as possible to 2212, because when the number of samples cannot be divided evenly by the number of cards, multi-card will fill the last batch to ensure each card has the same number of samples, affecting gradient synchronization.
```

Thanks!

**Comment by yifan123** [comment]
Setting config.sample.num_batches_per_epoch = 8 might be better. Additionally, I recently found that clip_range=1e-5 works better. In my previous experiments, 3k steps should have already reached 90. I’ll run the open-source code again on my end. Please wait a moment.

**Comment by yifan123** [comment]
Please use the following settings for 3 nodes:

```yaml
config.resolution = 512
config.sample.train_batch_size = 6
config.sample.num_image_per_prompt = 24
config.sample.num_batches_per_epoch = 8
config.train.beta = 0.004
```

The following are the results I reran over the past few days.

![Image](https://github.com/user-attachments/assets/afdd7497-199d-4a2b-8de5-2e4c3c14954e)

In practice, the main factor affecting training speed is the KL. You can check whether your KL is set larger than mine. If you set [[this line](https://github.com/yifan123/flow_grpo/blob/main/config/dgx.py#L107)](https://github.com/yifan123/flow_grpo/blob/main/config/dgx.py#L107) to a smaller value, you’ll notice that the training speed improves significantly, but the image quality might degrade more.

Additionally, because SD3.5 uses a scheduler where `dt` increases as you get closer to the final image, the later timesteps contribute less to the KL. Since our ultimate goal is to maintain image quality, the contribution of the later timesteps to the KL should actually be larger. You can remove the `dt` in the denominator when computing the KL. 

Specifically, when `num_timesteps = 10`, you can change the original `config.train.beta = 0.004` to `config.train.beta = 0.04` and remove the `dt` in the denominator when computing the KL (change `std_dev_t * torch.sqrt(-1*dt)` to `std_dev_t` in [line](https://github.com/yifan123/flow_grpo/blob/main/flow_grpo/diffusers_patch/sd3_sde_with_logprob.py#L80)), which reduces the drop in image quality while allowing the reward to increase faster.

![Image](https://github.com/user-attachments/assets/45bb59bc-b900-4f48-ac93-e05e3dfe05ad)

![Image](https://github.com/user-attachments/assets/2c620679-4101-44e9-bf3e-08b7c7b81521)

**Comment by yifan123** [comment]
Thanks for your interest in our work!

First, I think you can turn off the KL term — although this may lead to reward hacking, it will significantly speed up training. If you want to test a new RL optimization, you can disable the KL term first and add it back later after validation.

![Image](https://github.com/user-attachments/assets/65ad26eb-152c-492c-b686-895ffe449397)
In this figure, the red line shows the result without KL, and you can see that the reward improves right from the start. Please use the hyperparameters I provided, especially `config.sample.num_image_per_prompt=6` is too small — it will cause the estimated advantage to have very high variance. I recommend using `config.sample.num_image_per_prompt=24`.

Recommended settings:

```
# gpu_number = 4
# bs = 6
# group = 24
config.sample.train_batch_size = bs  # 6
config.sample.num_image_per_prompt = group  # 24
config.sample.num_batches_per_epoch = 48 / (bs * gpu_number / group)  # 48
config.train.beta = 0
```

**Comment by STAR-CV** [comment]
> Please use the following settings for 3 nodes:
> 
> config.resolution = 512
> config.sample.train_batch_size = 6
> config.sample.num_image_per_prompt = 24
> config.sample.num_batches_per_epoch = 8
> config.train.beta = 0.004
> The following are the results I reran over the past few days.
> 
> ![Image](https://github.com/user-attachments/assets/afdd7497-199d-4a2b-8de5-2e4c3c14954e)
> 
> In practice, the main factor affecting training speed is the KL. You can check whether your KL is set larger than mine. If you set [[this line](https://github.com/yifan123/flow_grpo/blob/main/config/dgx.py#L107)](https://github.com/yifan123/flow_grpo/blob/main/config/dgx.py#L107) to a smaller value, you’ll notice that the training speed improves significantly, but the image quality might degrade more.
> 
> Additionally, because SD3.5 uses a scheduler where `dt` increases as you get closer to the final image, the later timesteps contribute less to the KL. Since our ultimate goal is to maintain image quality, the contribution of the later timesteps to the KL should actually be larger. You can remove the `dt` in the denominator when computing the KL.
> 
> Specifically, when `num_timesteps = 10`, you can change the original `config.train.beta = 0.004` to `config.train.beta = 0.04` and remove the `dt` in the denominator when computing the KL (change `std_dev_t * torch.sqrt(-1*dt)` to `std_dev_t` in [line](https://github.com/yifan123/flow_grpo/blob/main/flow_grpo/diffusers_patch/sd3_sde_with_logprob.py#L80)), which reduces the drop in image quality while allowing the reward to increase faster.
> 
> ![Image](https://github.com/user-attachments/assets/45bb59bc-b900-4f48-ac93-e05e3dfe05ad)
> 
> ![Image](https://github.com/user-attachments/assets/2c620679-4101-44e9-bf3e-08b7c7b81521)

Hi, I would like to reproduce the results of ocr and pickscore from the paper, is it also trained according to this configuration? Or according to the corresponding configuration in dgx.py?

**Comment by yifan123** [comment]
Yes, you can directly use the corresponding configuration in `dgx.py`. 

Additionally, I actually recommend changing `config.train.beta` from `0.004` to `0.04` and removing the `dt` in the denominator when computing the KL term (specifically, change `std_dev_t * torch.sqrt(-1 * dt)` to `std_dev_t` in [[this line](https://github.com/yifan123/flow_grpo/blob/main/flow_grpo/diffusers_patch/sd3_sde_with_logprob.py#L80)](https://github.com/yifan123/flow_grpo/blob/main/flow_grpo/diffusers_patch/sd3_sde_with_logprob.py#L80)). This adjustment helps reduce the drop in image quality while allowing the reward to increase faster.

**Comment by silence917-byte** [comment]
Hi, great work, I would like to ask you about my reproduction of geneval w/o kl with the above config on 3 nodes, about 4000 steps, and I would like to ask why the reward_eval reaches about 0.92, but the reward_accuracy is at 0.85, and the metrics measured with general_multigpu are also at 0.85, mostly The position and colorattr performance is very poor, have you encountered this before, any training suggestions would be appreciated!

<img width="1295" height="348" alt="Image" src="https://github.com/user-attachments/assets/7fdd3a76-2aba-447d-8631-1a23515c4cb4" />
